pkg/apply/prune/prune.go

```diff
@@ -83,114 +83,174 @@ type Options struct {
 	// used for apply.
-	sort.Sort(sort.Reverse(ordering.SortableMetas(clusterObjs)))
-	for _, clusterObj := range clusterObjs {
-		mapping, err := po.mapper.RESTMapping(clusterObj.GroupKind)
-		if err != nil {
-			localIds = append(localIds, clusterObj)
-			taskContext.EventChannel() <- createPruneFailedEvent(clusterObj, err)
-			taskContext.CaptureResourceFailure(clusterObj)
-			continue
-		}
-		namespacedClient := po.client.Resource(mapping.Resource).Namespace(clusterObj.Namespace)
-		obj, err := namespacedClient.Get(context.TODO(), clusterObj.Name, metav1.GetOptions{})
+	sort.Sort(sort.Reverse(ordering.SortableMetas(pruneObjs)))
+	// Store prune failures to ensure they remain in the inventory.
+	pruneFailures := []object.ObjMetadata{}
+	for _, pruneObj := range pruneObjs {
+		klog.V(5).Infof("attempting prune: %s", pruneObj)
+		obj, err := po.getObject(pruneObj)
 		if err != nil {
 			// Object not found in cluster, so no need to delete it; skip to next object.
 			if apierrors.IsNotFound(err) {
+				klog.V(5).Infof("%s/%s not found in cluster--skipping",
+					pruneObj.Namespace, pruneObj.Name)
 				continue
 			}
-			localIds = append(localIds, clusterObj)
-			taskContext.EventChannel() <- createPruneFailedEvent(clusterObj, err)
-			taskContext.CaptureResourceFailure(clusterObj)
-			continue
-		}
-		metadata, err := meta.Accessor(obj)
-		if err != nil {
-			localIds = append(localIds, clusterObj)
-			taskContext.EventChannel() <- createPruneFailedEvent(clusterObj, err)
-			taskContext.CaptureResourceFailure(clusterObj)
+			if klog.V(5) {
+				klog.Errorf("prune obj (%s/%s) UID retrival error: %s",
+					pruneObj.Namespace, pruneObj.Name, err)
+			}
+			pruneFailures = append(pruneFailures, pruneObj)
+			taskContext.EventChannel() <- createPruneFailedEvent(pruneObj, err)
+			taskContext.CaptureResourceFailure(pruneObj)
 			continue
 		}
-		// If this cluster object is not also a currently applied
-		// object, then it has been omitted--prune it. If the cluster
-		// object is part of the local apply set, skip it.
-		uid := string(metadata.GetUID())
-		klog.V(7).Infof("prune previously applied object UID: %s", uid)
+		// Do not prune objects that are in set of currently applied objects.
+		uid := string(obj.GetUID())
 		if currentUIDs.Has(uid) {
-			klog.V(7).Infof("prune object in current apply; do not prune: %s", uid)
+			klog.V(5).Infof("prune object in current apply; do not prune: %s", uid)
 			continue
 		}
 		// Handle lifecycle directive preventing deletion.
 		if !canPrune(localInv, obj, o.InventoryPolicy, uid) {
-			taskContext.EventChannel() <- createPruneEvent(clusterObj, obj, event.PruneSkipped)
-			localIds = append(localIds, clusterObj)
+			klog.V(4).Infof("skip prune for lifecycle directive %s/%s", pruneObj.Namespace, pruneObj.Name)
+			taskContext.EventChannel() <- createPruneEvent(pruneObj, obj, event.PruneSkipped)
+			pruneFailures = append(pruneFailures, pruneObj)
```

Comment:
Maybe add a comment here for why adding it to `pruneFailures`. It's actually not a failure, we want to use `pruneFailures` to keep it in the final inventory list.